amethyst_ui/src/text.rs

```diff
@@ -0,0 +1,166 @@
+    font_size: f32,
+    /// This is true if the texture needs to be re-rendered
+    dirty: bool,
+}
+
+impl UiText {
+    /// Initializes a new UiText
+    ///
+    /// # Parameters
+    ///
+    /// * `font`: A handle to a `Font` asset
+    /// * `text`: the glyphs to render
+    /// * `color`: RGBA color with a maximum of 1.0 and a minimum of 0.0 for each channel
+    /// * `font_size`: a uniform scale applied to the glyphs
+    pub fn new(font: FontHandle, text: String, color: [f32; 4], font_size: f32) -> UiText {
+        UiText {
+            texture: None,
+            font,
+            text,
+            color,
+            font_size,
+            dirty: true,
+        }
+    }
+
+    /// Retrieves the handle to the font this is using.
+    pub fn font(&self) -> &FontHandle {
+        &self.font
+    }
+
+    /// Sets the font for this to use in rendering.
+    ///
+    /// Avoid calling this if you don't have to, as calls to this invalidate the text render cache.
+    pub fn set_font(&mut self, font: FontHandle) {
+        self.font = font;
+        self.dirty = true;
+    }
+
+    /// Returns the string slice used by this.
+    pub fn text(&self) -> &str {
+        &self.text
+    }
+
+    /// Gets a mutable reference to the string to render.
+    ///
+    /// Avoid calling this if you don't have to, as calls to this invalidate the text render cache.
+    pub fn text_mut(&mut self) -> &mut String {
+        self.dirty = true;
+        &mut self.text
+    }
+
+    /// The RGBA color with a maximum of 1.0 and a minimum of 0.0 for each channel of the text.
+    pub fn color(&self) -> [f32; 4] {
+        self.color
+    }
+
+    /// Set the RGBA color with a maximum of 1.0 and a minimum of 0.0 for each channel.
+    pub fn set_color(&mut self, color: [f32; 4]) {
+        self.color = color;
+        self.dirty = true;
+    }
+
+    /// The maximum height of the font in pixels.
+    pub fn font_size(&self) -> f32 {
+        self.font_size
+    }
+
+    /// Set the maximum height of the font in pixels.
+    ///
+    /// Avoid calling this if you don't have to, as calls to this invalidate the text render cache.
+    pub fn set_font_size(&mut self, size: f32) {
+        self.dirty = true;
+        self.font_size = size;
+    }
+}
+
+impl Component for UiText {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// This system renders `UiTex`t.
+/// Make sure it's called after changes are made to `UiText` but before
+/// the `RenderSystem` from `amethyst_renderer` gets invoked.
+pub struct UiTextRenderer;
+
+impl<'a> System<'a> for UiTextRenderer {
+    type SystemData = (
+        ReadStorage<'a, UiTransform>,
+        WriteStorage<'a, UiText>,
+        Fetch<'a, Loader>,
+        Fetch<'a, AssetStorage<Texture>>,
+        Fetch<'a, AssetStorage<FontAsset>>,
+    );
+
+    fn run(&mut self, (transform, mut text, loader, tex_storage, font_storage): Self::SystemData) {
+        for (transform, text) in (&transform, &mut text).join() {
```

Comment:
I'm not sure if this is a more general problem, but if a text doesn't have a transform, it will be silently ignored. No need to fix it here, just wanted to note.

Reply:
Please filter and flat_map this, otherwise the indention is too much.